### PR TITLE
Add a new function "try_get_value"

### PR DESCRIPTION
**Pull request by dmitryilyin**
Extracts a value from a deeply-nested data structure

$data = {
  'a' => {
    'b' => [
      'b1',
      'b2',
      'b3',
    ]
  }
}

$value = fetch($data, 'a/b/2')
=> $value = 'b3'

### CODE REVIEW

File: README.markdown [hunk 1]
```diff
@@ -1023,6 +1023,40 @@ Versions | Puppet 2.6 | Puppet 2.7 | Puppet 3.x | Puppet 4.x |
 
 **stdlib 5.x**: When released, stdlib 5.x will drop support for Puppet 2.7.x. Please see [this discussion](https://github.com/puppetlabs/puppetlabs-stdlib/pull/176#issuecomment-30251414).
 
+#### `fetch`
```
Comment (DavidS) [inline README.markdown]: Please put this in the right alphabetical order of the function list.

File: lib/puppet/parser/functions/fetch.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
+module Puppet::Parser::Functions
+  newfunction(
+      :fetch,
```
Comment (DavidS) [inline lib/puppet/parser/functions/fetch.rb]: 'fetch' is a very generic name. It's probably a good investment in the future to think hard about a more specific name.

File: lib/puppet/parser/functions/fetch.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    path_lookup = lambda do |data, path, default|
+      # no data -> return default
+      break default if data.nil?
+      # wrong path -> return default
+      break default unless path.is_a? Array
+      # empty path -> value found
+      break data unless path.any?
+      # non-empty path and non-structure data -> return default
+      break default unless data.is_a? Hash or data.is_a? Array
```
Comment (DavidS) [inline lib/puppet/parser/functions/fetch.rb]: All of these conditions create a complex beast. Having a debug output specifying what is being looked up and which decisions were made would greatly help users of this.

File: lib/puppet/parser/functions/fetch.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+      key = path.shift
+      if data.is_a? Array
+        begin
+          key = Integer key
```
Comment (DavidS) [inline lib/puppet/parser/functions/fetch.rb]: `key.to_i` would look more ruby-ish.

File: lib/puppet/parser/functions/fetch.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      path_lookup.call data[key], path, default
+    end
+
+    data = args[0]
+    path = args[1]
+    default = args[2]
+    separator = args[3]
+    separator = '/' unless separator
+
+    path = '' unless path
```
Comment (DavidS) [inline lib/puppet/parser/functions/fetch.rb]: Use `path = args[1] || ''` for more ruby-ishness.

File: spec/acceptance/fetch_spec.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      $data = {
+        'a' => { 'b' => 'passing'}
+      }
+
+      $tests = fetch($a, 'a/b')
+      notice(inline_template('tests are <%= @tests.inspect %>'))
+      EOS
+
+      apply_manifest(pp, :catch_failures => true) do |r|
+        expect(r.stdout).to match(/tests are "passing"/)
```
Comment (DavidS) [inline spec/acceptance/fetch_spec.rb]: brilliant use of test data :-D

File: spec/functions/fetch_spec.rb [hunk 7]
```diff
@@ -0,0 +1,136 @@
+require 'spec_helper'
+
```
Comment (DavidS) [inline spec/functions/fetch_spec.rb]: Please restructure this to match the current pattern of unit tests in stdlib.

Using the `run.with_params(...).and_return(...)` matcher makes for so much nicer reading of the code.

**Comment by DavidS** [comment]
Hi, thank you for your contribution! I've added a few inline tests about possible improvements for the code.

Cheers, David

**Comment by dmitryilyin** [comment]
Updated request with fixes

File: spec/functions/getpath_spec.rb [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    end
+
+    it 'should return the default value if the path is too long' do
+      data = {
+          'a' => {
+              'b' => 'c'
+          }
+      }
+      is_expected.to run.with_params(data, 'a/b/c/d', 'default').and_return('default')
+    end
```
Comment (DavidS) [inline spec/functions/getpath_spec.rb]: You can shorten your tests even more by sharing data between examples:

```
context 'with nested hashes' do
  let(:data) { { 'a' => { 'b' => 'c' } } }
  it 'should return the default value if the path is not found' do
    is_expected.to run.with_params(data, 'missing', 'default').and_return('default')
  end
  it 'should return the default value if the path is too long' do
    is_expected.to run.with_params(data, 'a/b/c/d', 'default').and_return('default')
  end
end
```

**Comment by DavidS** [comment]
These changes look great. I've added another small comment to the tests.

We've been discussing names internally and came up with `try_get_value` to avoid confusion with filesystem paths and to indicate that it will not throw any exceptions.

**Comment by DavidS** [comment]
Thank you for your contribution and patience!

**Pull request merged**